index.js

```diff
@@ -63,24 +43,20 @@ app.get('/product/:id', (req, res) => {
     productId = `P${userInput.padStart(3, '0')}`
   }
 
-  for (let product of productData.products) {
-    if (product.id === productId) {
-      jsonOutputObject = product;
-      break;
-    }
-  }
+  // Creating required Maps
+  dataService.getProducts;
+  dataService.getCategories;
+
+  jsonOutputObject = dataService.productMap[productId];
 
-  // Checking for null at the end would cause NullPointerException when trying to get the categoryName
+  // Checking for null after adding categoryName causes errors
   if (jsonOutputObject == null) {
     res.send('Product not found!');
     return;
   }
 
-  for (let category of categoryData.categories) {
-    if (jsonOutputObject.categoryId === category.id) {
-      jsonOutputObject.categoryName = category.categoryName;
-    }
-  }
+  // Adding categoryName
+  jsonOutputObject.categoryName = dataService.categoryMap[jsonOutputObject.categoryId];
```

Comment:
Get from `getCombinedProductMap` instead, so business logic like getting `categoryName` is not done here.

Reply:
Yes, even I wanted to do that this way but then I taught `getProducts()` function would become useless. Anyway Done!